ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/manager/EthPeers.java

```diff
@@ -102,48 +125,98 @@ public EthPeers(
         pendingRequests::size);
+    metricsSystem.createIntegerGauge(
+        BesuMetricCategory.ETHEREUM,
+        "peer_limit",
+        "The maximum number of peers this node allows to connect",
+        () -> peerUpperBound);
+    connectedPeersCounter =
+        metricsSystem.createCounter(
+            BesuMetricCategory.PEERS, "connected_total", "Total number of peers connected");
   }
 
-  public void registerConnection(
-      final PeerConnection peerConnection, final List<PeerValidator> peerValidators) {
-    final EthPeer peer =
-        new EthPeer(
-            peerConnection,
-            protocolName,
-            this::invokeConnectionCallbacks,
-            peerValidators,
-            maxMessageSize,
-            clock,
-            permissioningProviders);
-    connections.putIfAbsent(peerConnection, peer);
-    LOG.debug("Adding new EthPeer {}", peer.nodeId());
-  }
-
-  public void registerDisconnect(final PeerConnection connection) {
-    final EthPeer peer = connections.remove(connection);
-    if (peer != null) {
-      disconnectCallbacks.forEach(callback -> callback.onDisconnect(peer));
-      peer.handleDisconnect();
-      abortPendingRequestsAssignedToDisconnectedPeers();
-      LOG.debug("Disconnected EthPeer {}", peer);
+  public void registerNewConnection(
+      final PeerConnection newConnection, final List<PeerValidator> peerValidators) {
+    final Bytes id = newConnection.getPeer().getId();
+    synchronized (this) {
+      EthPeer ethPeer = connections.get(id);
+      if (ethPeer == null) {
+        final Optional<EthPeer> peerInList =
+            notReadyConnections.asMap().values().stream()
+                .filter(p -> p.getId().equals(id))
+                .findFirst();
+        ethPeer =
+            peerInList.orElse(
+                new EthPeer(
+                    newConnection,
+                    protocolName,
+                    this::ethPeerStatusExchanged,
+                    peerValidators,
+                    maxMessageSize,
+                    clock,
+                    permissioningProviders,
+                    localNodeId));
+      }
+      notReadyConnections.put(newConnection, ethPeer);
+    }
+  }
+
+  @NotNull
+  private List<PeerConnection> getNotReadyConnections(final Bytes id) {
+    return notReadyConnections.asMap().keySet().stream()
+        .filter(nrc -> nrc.getPeer().getId().equals(id))
+        .collect(Collectors.toList());
+  }
+
+  public boolean registerDisconnect(final PeerConnection connection) {
+    final Bytes id = connection.getPeer().getId();
+    final EthPeer peer = connections.get(id);
+    return registerDisconnect(id, peer, connection);
+  }
+
+  private boolean registerDisconnect(
+      final Bytes id, final EthPeer peer, final PeerConnection connection) {
+    notReadyConnections.invalidate(connection);
+    boolean removed = false;
+    if (peer != null && peer.getConnection().equals(connection)) {
+      if (!peerHasNonReadyConnection(id)) {
```

Comment:
is there a better name than "non ready" - incomplete? partial?

Reply:
chaged it to incomplete! Far better!